File: packages/core/src/event-bus.ts
import type { SteadwingEvent } from "./types.js";

export type EventListener = (event: SteadwingEvent) => void;

/**
 * Simple pub/sub event bus for SteadwingEvents.
 * Replaces framework-specific event subscriptions (e.g., OpenClaw's onDiagnosticEvent).
 */
export class SteadwingEventBus {
  private listeners = new Set<EventListener>();

  /** Subscribe to all events. Returns an unsubscribe function. */
  on(listener: EventListener): () => void {
    this.listeners.add(listener);
    return () => { this.listeners.delete(listener); };
  }

  /** Emit an event to all listeners. Errors in listeners are caught and logged. */
  emit(event: SteadwingEvent): void {
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch (err) {
        console.error("[steadwing] event listener error:", err);
      }
    }
  }

  /** Remove all listeners. */
  clear(): void {
    this.listeners.clear();
  }

  /** Current listener count. */
  get size(): number {
    return this.listeners.size;
  }
}

File: python/src/core/event_bus.py
from __future__ import annotations

import logging
from collections.abc import Awaitable, Callable

from openalerts.core.types import OpenAlertsEvent

logger = logging.getLogger("openalerts.event_bus")

Listener = Callable[[OpenAlertsEvent], Awaitable[None]]


class OpenAlertsEventBus:
    """Async pub/sub event bus."""

    def __init__(self) -> None:
        self._listeners: set[Listener] = set()

    def on(self, listener: Listener) -> Callable[[], None]:
        """Subscribe a listener. Returns an unsubscribe callable."""
        self._listeners.add(listener)

        def unsubscribe() -> None:
            self._listeners.discard(listener)

        return unsubscribe

    async def emit(self, event: OpenAlertsEvent) -> None:
        """Broadcast event to all listeners. Errors caught per-listener."""
        for listener in list(self._listeners):
            try:
                await listener(event)
            except Exception:
                logger.exception("Listener %s failed for event %s", listener, event.type)

    def clear(self) -> None:
        self._listeners.clear()

    @property
    def size(self) -> int:
        return len(self._listeners)
